packages/network-controller/src/NetworkController.ts

```diff
@@ -1941,6 +1958,129 @@ export class NetworkController extends BaseController<
+        (rpcEndpoint) => {
+          return {
+            type: 'add' as const,
+            rpcEndpoint,
+          };
+        },
+      );
+      this.#registerNetworkClientsAsNeeded({
+        networkFields: networkConfiguration,
+        networkClientOperations: networkClientAddOperations,
+        autoManagedNetworkClientRegistry,
+      });
+    };
+
+    // Replace the networkConfiguration with our new networkConfiguration
+    // This is a full replace (no merging)
+    const replaceNetworkConfiguration = () => {
+      // Update State
+      this.update((state) => {
+        state.networkConfigurationsByChainId[networkConfiguration.chainId] =
+          networkConfiguration;
+      });
+
+      // Update Cache
+      this.#networkConfigurationsByNetworkClientId =
+        buildNetworkConfigurationsByNetworkClientId(
+          this.state.networkConfigurationsByChainId,
+        );
+    };
+
+    // Updates the NetworksMetadata State
+    const updateNetworksMetadata = async () => {
+      // Remove old metadata state
+      this.update((state) => {
+        prevNetworkConfig.rpcEndpoints.forEach((r) => {
+          if (state.networksMetadata?.[r.networkClientId]) {
+            delete state.networksMetadata[r.networkClientId];
+          }
+        });
+      });
+
+      // Add new metadata state
+      for (const r of networkConfiguration.rpcEndpoints) {
+        await this.lookupNetwork(r.networkClientId);
+      }
+    };
+
+    // Update selectedNetworkId State
+    // Will try to keep the same OR will select a new RPC from new network OR any network (edge case)
+    const updateSelectedNetworkId = async () => {
+      const selectedClientId = this.state.selectedNetworkClientId;
+      const wasClientIdReplaced = prevNetworkConfig.rpcEndpoints.some(
+        (r) => r.networkClientId === selectedClientId,
+      );
+      const doesExistInNewNetwork = networkConfiguration.rpcEndpoints.some(
+        (r) => r.networkClientId === selectedClientId,
+      );
+
+      const shouldUpdateSelectedNetworkId =
+        wasClientIdReplaced && !doesExistInNewNetwork;
+      if (shouldUpdateSelectedNetworkId) {
+        // Update the clientId to "something" that exists
+        const newRPCClientId = networkConfiguration.rpcEndpoints.find(
+          (r) => r.networkClientId in this.state.networksMetadata,
+        )?.networkClientId;
+        const anyRPCClientId = Object.keys(this.state.networksMetadata)[0];
+        /* istanbul ignore next: anyRPCClientId and selectedClientId are fallbacks and should be impossible to reach  */
+        const newlySelectedNetwork =
+          newRPCClientId ?? anyRPCClientId ?? selectedClientId;
+        await this.#refreshNetwork(newlySelectedNetwork);
+      }
+    };
+
+    // Execute Set Network Config
+    updateRegistry();
+    replaceNetworkConfiguration();
+    await updateNetworksMetadata();
+    await updateSelectedNetworkId();
```

Comment:
Looks pretty solid!

Reply:
Haha, the network controller logic is _a lot_. Even though these functions are only called once, I wanted to keep the steps distinct.

In the future, it would be cool moving a lot of this logic out of the controller. However there are many inter-dependencies to controller properties/methods. It would be more cumbersome to extract this out with a lot of parameters to pass in.